### PR TITLE
[WIP] Adds SQLAlchemy code for QuestionAnswerPairs

### LINKED ISSUE #52: Create an SQLAlchemy object to store QuestionAnswerPairs
## Objective
Create an SQLAlchemy object to store QuestionAnswerPairs

## Key Result
- [ ] Commit code to the `Entities` folder that defines the `QuestionAnswerPairs` object 
- [ ] Commit code to the `database_wrapper.py` that allows `save_qa_pair` and `get_all_qa_pairs`

### PR DESCRIPTION
**Pull request by mfekadu**
## What's New?
* adds create & update functionality for Question & Answer data for the nimbus NLP pipeline

## Type of change (pick-one)
- [ ] Bug fix (_non-breaking change which fixes an issue_)
- [x] New feature (_non-breaking change which adds functionality_)
- [ ] Breaking change (_fix or feature that would cause existing functionality to not work as expected_)
- [ ] This change requires a documentation update

## How Has This Been Tested?
Yes see comment below

## Checklist (check-all-before-merge)
_formatting help: `- [x]` means "checked' and `- [ ]` means "unchecked"_

- [ ] I documented my code according to the [Google Python Style Guide][1]

- [ ] I ran `./build_docs.sh` and the docs look fine

- [ ] I ran `./type_check.sh` and got no errors

- [ ] I ran `./format.sh` because it automatically cleans my code for me 😄 

- [ ] I ran `./lint.sh` to check for what "format" missed

- [ ] I added my tests to the `/tests` directory

- [ ] I ran `./run_tests.sh` and all the tests pass

[1]: https://google.github.io/styleguide/pyguide.html

### CODE REVIEW

File: Entity/QuestionAnswerPair.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    question_format = Column(Text)
+    answer_format = Column(Text)
+
+    def __repr__(self):
+        """
+        A lazy __repr__ inspired by https://stackoverflow.com/a/60087190
+        """
+        D = self.__dict__
+        attributes = [
+            f"{k}={D.get(k)}" for k in self.__dir__() if not k.startswith("_")
```
Comment (snekiam) [inline Entity/QuestionAnswerPair.py:36]: Can you add a comment about what this line does? Even something as simple as "This creates a list of attributes that don't start with an '_'." could be helpful in the future I think.

**Comment by mfekadu** [comment]
# What's new
## Made the entity
https://github.com/calpoly-csai/api/blob/11b6ecd0f8b720dac15067201f999498037f3d2c/Entity/QuestionAnswerPair.py#L1-L39

## Modified the `database_wrapper`
### import the entity
https://github.com/calpoly-csai/api/blob/11b6ecd0f8b720dac15067201f999498037f3d2c/database_wrapper.py#L23

### add `self.QuestionAnswerPair`
https://github.com/calpoly-csai/api/blob/11b6ecd0f8b720dac15067201f999498037f3d2c/database_wrapper.py#L253-L262

### do `__safe_create(self.QuestionAnswerPair)`
https://github.com/calpoly-csai/api/blob/11b6ecd0f8b720dac15067201f999498037f3d2c/database_wrapper.py#L290-L305

### define `save_question_answer_pair`
https://github.com/calpoly-csai/api/blob/11b6ecd0f8b720dac15067201f999498037f3d2c/database_wrapper.py#L332-L364

File: database_wrapper.py [hunk 2]
```diff
@@ -247,72 +250,76 @@ def wrapper(*args, **kwargs):
...
         print("initialized NimbusMySQLAlchemy")
 
     def _create_all_tables(self):
-        # TODO: reconsider if this even works???
-        # self.Base.metadata.create_all(self.engine)
-        # TODO: go with this instead... more explicit...
         def __safe_create(SQLAlchemy_object):
             table_name = SQLAlchemy_object.__tablename__
-            print("creating {}...".format(table_name))
+            print(f"creating {table_name}...")
```
Comment (snekiam) [inline database_wrapper.py:296]: I'm not familiar with this syntax - is this how we want to do string formatting?

File: database_wrapper.py [hunk 3]
```diff
@@ -322,12 +329,46 @@ def create_AudioSampleMetaData_table(self) -> None:
...
+        Args:
+            qa_dict: a dictionary that corresponds to the fields in QuestionAnswerPair  # noqa
+
+        Raises:
+            BadDictionaryKeyError - ...
+            BadDictionaryValueError - ...
+
+        Returns:
+            True if all is good, else False
+        """
```
Comment (snekiam) [inline database_wrapper.py:356]: I love the long description of inputs and outputs! Should be very helpful for future maintenance.

File: database_wrapper.py [hunk 4]
```diff
@@ -377,48 +426,54 @@ def save_audio_sample_meta_data(self, formatted_data: dict) -> bool:
...
+            formatted_data["noiseLevel"] == "m"
+            or formatted_data["noiseLevel"] == "medium"
+        ):
             metadata.noise_level = NoiseLevel.medium
-        elif (formatted_data['noiseLevel'] == 'l' or
-              formatted_data['noiseLevel'] == 'loud'):
+        elif (
+            formatted_data["noiseLevel"] == "l"
+            or formatted_data["noiseLevel"] == "loud"
+        ):
```
Comment (snekiam) [inline database_wrapper.py:460]: I'm not sure if I like the multiline if statements - it seems like it takes up a lot of vertical space (I know this was the linter - maybe we should change some settings in the future? I don't think this should block the PR.)

File: database_wrapper.py [hunk 5]
```diff
@@ -607,10 +666,9 @@ def get_course_properties(self, courseName) -> List[str]:
         """
 
         # FIXME: resolve unused variable `props`, until then, commented out
-        props = self.get_property_from_entity(["*"],
-                                              "Courses",
-                                              condition_field="courseName",
-                                              condition_value=courseName)
+        props = self.get_property_from_entity(
+            ["*"], "Courses", condition_field="courseName", condition_value=courseName
+        )
```
Comment (snekiam) [inline database_wrapper.py:713]: Should this be commented out as indicated in the FIXME?

**Review by snekiam — APPROVED** [review]
Looks good, maybe I need to read some linting docs though - I'm not familiar with the linter and some of it is probably best practice.

**Comment by mfekadu** [comment]
# How Has This Been Tested?

Feb-23-2020 14-27-32

File: database_wrapper.py (hunk 2 above)
Comment (mfekadu) [reply database_wrapper.py:296]: Yes, `f-strings` are a thing https://realpython.com/python-f-strings/

They are super cool!

Thanks for your curiosity @snekiam

File: database_wrapper.py (hunk 5 above)
Comment (mfekadu) [reply database_wrapper.py:713]: @snekiam  good point, I'll make another PR of refactorings for dead code, FIXMEs & TODOs

File: database_wrapper.py (hunk 4 above)
Comment (mfekadu) [reply database_wrapper.py:460]: https://black.readthedocs.io

File: database_wrapper.py (hunk 3 above)
Comment (mfekadu) [reply database_wrapper.py:356]: ❤️

File: Entity/QuestionAnswerPair.py (hunk 1 above)
Comment (mfekadu) [reply Entity/QuestionAnswerPair.py:36]: @snekiam  good point, I'll make another PR of refactorings for dead code, FIXMEs & TODOs

**Pull request merged**